### DISCUSSION TITLE
Visual stimuli and serial writing not in sync

### DISCUSSION THREAD

**Discussion by czhao356** [bonsai-rx/bonsai#1697] [Q&A] [answered]
Hello all,

I made this protocol that sends the number of light pulses to a LED driver and draws a visual grating stimuli at the same time. However based on our recording it seems like the LED pulses are always leading the visual stimuli by 70-90ms. I'm new to Bonsai so I'm not sure why I can't get everything on the same clock. 

The bonsai files are attached. I also included a screenshot of our recording system, showing the LED pulse (ADC4 channel, from the LED driver) comes up earlier than the visual stimuli (ADC5 channel, recorded by a photodiode). We use OpenEphys system and the LED driver is the Cyclops from OpenEphys. Both the LED driver output and photodiode output are plugged into the same I/O board.

Thanks!

![recording](https://github.com/bonsai-rx/bonsai/assets/88522738/c4f088c1-d4fc-46f8-b589-910e0c3848a4)
![bonsai](https://github.com/bonsai-rx/bonsai/assets/88522738/3e952864-149e-433a-a8a1-899c638f295e)

[Feb272024_optostim_with_grating.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/14452284/Feb272024_optostim_with_grating.bonsai.zip)

**Comment by bruno-f-cruz** [comment]
If you want precise alignment the easiest would be to trigger the photodiode using the output of the photodiode, no?

**Reply by czhao356** [reply]
Yes I'm thinking about doing that. I'm just trying to combine 2 established protocols (one presents visual stimuli, the other sends light pulses through the LED driver) that has been in use in my lab. I'm not very familiar with arduino so I'm trying not to modify the LED driver program. We have to change the number of light pulses and voltage every second, so I think reading a csv file is still the easiet way?

Anyways rignt now I'm trying to figure out if the timing difference is because I made a mistake in the Bonsai workflow.

**Reply by bruno-f-cruz** [reply, marked as answer]
If I understand correctly, you are asking if you can concurrently trigger, in the real world, two completely different software events. The answer is "no" unfort. One of the events (rendering) is subject to the latencies of the rendering pipeline. These latencies are quite variable but usually within 2-5 frames. So if you have a 60hz monitor you should be looking at ~30-80ms. The serial communication latencies are quite smaller (<10ms) depending on the drivers you are using. I think this likely explains the kind of numbers you are seeing in your oscilloscope.

If you want to get the two phenomena to occur "at the same time", the best solution would be what I suggested previously where you laser would be triggered by a physical change of the monitor output.
